pkg/codegen/go/gen_program_optionals.go

```diff
@@ -30,91 +28,137 @@ func (ot *optionalTemp) SyntaxNode() hclsyntax.Node {
-								if strKey == op {
-									temp = &optionalTemp{
-										Name:  fmt.Sprintf("opt%d", os.count),
-										Value: item.Value,
-									}
-									os.temps = append(os.temps, temp)
-									os.count++
-									x.Items[i].Value = &model.ScopeTraversalExpression{
-										RootName:  fmt.Sprintf("&%s", temp.Name),
-										Traversal: hcl.Traversal{hcl.TraverseRoot{Name: ""}},
-										Parts:     []model.Traversable{temp},
-									}
+							if schemaType, isOptional := optionalPrimitives[strKey]; isOptional {
+								functionName := os.getOptionalConversion(schemaType)
+								expectedModelType := os.getExpectedModelType(schemaType)
+
+								x.Items[i].Value = &model.FunctionCallExpression{
+									Name: functionName,
+									Signature: model.StaticFunctionSignature{
+										Parameters: []model.Parameter{{
+											Name: "val",
+											Type: expectedModelType,
+										}},
+										ReturnType: model.NewOptionalType(expectedModelType),
+									},
+									Args: []model.Expression{item.Value},
 								}
 							}
 						}
 					}
 				}
 			}
 		}
+		// Clear before visiting children, require another __convert call to set again
+		os.intrinsicConvertTo = nil
+		return x, nil
+	default:
+		// Ditto
+		os.intrinsicConvertTo = nil
+		return x, nil
+	}
+	return x, nil
+}
+
+func (os *optionalSpiller) postVisitor(x model.Expression) (model.Expression, hcl.Diagnostics) {
+	switch x := x.(type) {
+	case *model.FunctionCallExpression:
+		if x.Name == "invoke" {
+			if x == os.invocation {
+				// Clear invocation flag once we're done traversing children.
+				os.invocation = nil
+			}
+		}
 	}
 	return x, nil
 }
 
-func (os *optionalSpiller) spillExpression(x model.Expression) (model.Expression, hcl.Diagnostics) {
-	isInvoke := false
-	return os.spillExpressionHelper(x, x.Type(), isInvoke)
+func (*optionalSpiller) getOptionalConversion(ty schema.Type) string {
+	switch ty {
+	case schema.NumberType:
+		return "goOptionalFloat64"
+	case schema.BoolType:
+		return "goOptionalBool"
+	case schema.IntType:
+		return "goOptionalInt"
+	case schema.StringType:
+		return "goOptionalString"
+	default:
+		return ""
+	}
+}
+
+func (*optionalSpiller) getExpectedModelType(ty schema.Type) model.Type {
+	switch ty {
+	case schema.NumberType:
+		return model.NumberType
+	case schema.BoolType:
+		return model.BoolType
+	case schema.IntType:
+		return model.IntType
+	case schema.StringType:
+		return model.StringType
+	default:
+		return nil
+	}
 }
 
 func (g *generator) rewriteOptionals(
 	x model.Expression,
 	spiller *optionalSpiller,
 ) (model.Expression, []*optionalTemp, hcl.Diagnostics) {
-	spiller.temps = nil
-	x, diags := model.VisitExpression(x, spiller.spillExpression, nil)
-
-	return x, spiller.temps, diags
+	// We want to recurse but we only want to use the previsitor, if post visitor is nil we don't
+	// recurse.
+	x, diags := model.VisitExpression(x, spiller.preVisitor, spiller.postVisitor)
```

Comment:
There is something really confusing here that changing postVisitor from nil changes the traversal semantics of VisitExpression.

What we had before:

```
  x, diags := model.VisitExpression(x, spiller.spillExpression, nil)
```

Was exactly equivalent to:

```
x, diags := spiller.spillExpression(x)
```

So there was no recursion whatsoever.

What we have now with post != nil, is that `model.VisitExpression` will visit every sub-expression.

Is this intentional?

Reply:
That's correct, without the recursion we don't get into the case in the test, we need to go a few levels deeper to get into the nested __convert